### ISSUE TITLE
Make it easy to wrap up programs

### ISSUE THREAD

**Issue by pelson** [pelson/conda-execute#20]
@msarahan was describing an issue where he wanted a python2 available in an environment in which it couldn't be supported (with a newer MSVC). In actual fact, conda provides a barrier to prevent people from installation of packages which cannot talk to one another inter-process - but sometimes what we really want is for the communication to happen across distinct processes, where no such restriction is needed.

What would be awesome (if it isn't already possible), is to be able to define a conda-execute script which delegates all arguments through to a wrapped executable in a different environment.

**Comment by pelson** [comment]
Pretty magically, this is already perfectly achievable:

```
#!/usr/bin/env conda-execute
"""
A wrapper script to provide python from a distinct environment.

"""
# conda execute
# env:
#  - python 2.*
# run_with: python 

import subprocess
import sys

cmd = [sys.executable] + sys.argv[1:]
subprocess.Popen(cmd).communicate()
```


@msarahan - I'd say your usecase can pretty much be solved by using this technique (though I confess I haven't tested features in earnest with conda-execute).

**Comment by msarahan** [comment]
so where and when would this conda execute script get written?  Is it part of the recipe?  Is there a magic syntax we can add to meta.yaml, like:

```
requirements:
  build:
    - execute:python=2.7
```

Not right now, but could there be?  The name of the env doesn't matter - so we can dynamically generate unique envs (with uuids?) for each execute dependency.

Also, what I would need to provide in my case is not just a way to run a command with an external environment, but to make that way to run a command available to some build process, like make or cmake.  In my case, I am not running python myself, but rather qt5's build process is.

**Comment by pelson** [comment]
> Is there a magic syntax we can add to meta.yaml

Really interesting example. I love the idea of having that option - you are expressing the dependency at a process level rather than intra-process.

> so we can dynamically generate unique envs (with uuids?) for each execute dependency.

conda-execute is doing precisely that. The UUID is determined by the sha of the environment specification.

> Also, what I would need to provide in my case is not just a way to run a command with an external environment

With existing conda-build, you could just put the code above into ```bin/python``` in your environment, make sure conda-execute is installed (as a build dep perhaps) and *voila*.

I just threw this recipe together:

```
pelson@~/dev/conda-execute/delme> cat meta.yaml 
package:
  name: foo
  version: 1

requirements:
  build:
    - conda-execute
    - python 3*

```

```
pelson@~/dev/conda-execute/delme> cat build.sh 
mkdir $PREFIX/python2
cp $RECIPE_DIR/python2 $PREFIX/bin/python2

which python
python --version

which python2
python2 --version
```

```
pelson@~/dev/conda-execute/delme> cat python2 
#!/usr/bin/env conda-execute
"""
A wrapper script to provide python from a distinct environment.

"""
# conda execute
# env:
#  - python 2.*
# run_with: python 

import subprocess
import sys

cmd = [sys.executable] + sys.argv[1:]
subprocess.Popen(cmd).communicate()
```

It runs quite slowly (because currently, conda-execute is making its own pkg cache and having to download the full py2k stack each build), but does work as expected (well, apart from the fact that conda-build thinks the pkg cache needs packaging up, which it clearly shouldn't be). Still a few warts, but proves the concept, I think.

**Comment by msarahan** [comment]
Yes, this is very nice.  If you like the idea of changing meta.yaml to support that feature, let's work towards that as a goal.  You'll notice I've created a conda-build 2.0 milestone.  Let's aim for that.

**Comment by pelson** [comment]
If this were to go into conda-build, it would also be valuable to have it in conda. I know @tkluyver is really interested in the ability to have python2 and python side-by-side in this way (so that Jupyter can be py3 only, but still provide a kernel for py2).

**Comment by msarahan** [comment]
Yes, I think you can't have one without the other, with conda being actually more important than conda-build in supporting this.  I'll talk with @kalefranz about it.
